**Data Preprocessing**

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Analyzing-borrowers’-risk-of-defaulting" data-toc-modified-id="Analyzing-borrowers’-risk-of-defaulting-1">Analyzing borrowers’ risk of defaulting</a></span><ul class="toc-item"><li><span><a href="#Step-1.-Open-the-data-file-and-have-a-look-at-the-general-information." data-toc-modified-id="Step-1.-Open-the-data-file-and-have-a-look-at-the-general-information.-1.1">Step 1. Open the data file and have a look at the general information.</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.2">Conclusion</a></span></li><li><span><a href="#Step-2.-Data-preprocessing" data-toc-modified-id="Step-2.-Data-preprocessing-1.3">Step 2. Data preprocessing</a></span></li><li><span><a href="#Processing-missing-values" data-toc-modified-id="Processing-missing-values-1.4">Processing missing values</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.5">Conclusion</a></span></li><li><span><a href="#Data-type-replacement" data-toc-modified-id="Data-type-replacement-1.6">Data type replacement</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.7">Conclusion</a></span></li><li><span><a href="#Processing-duplicates" data-toc-modified-id="Processing-duplicates-1.8">Processing duplicates</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.9">Conclusion</a></span></li><li><span><a href="#Categorizing-Data" data-toc-modified-id="Categorizing-Data-1.10">Categorizing Data</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.11">Conclusion</a></span></li><li><span><a href="#Step-3.-Answer-these-questions" data-toc-modified-id="Step-3.-Answer-these-questions-1.12">Step 3. Answer these questions</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.13">Conclusion</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.14">Conclusion</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.15">Conclusion</a></span></li><li><span><a href="#Step-4.-General-conclusion" data-toc-modified-id="Step-4.-General-conclusion-1.16">Step 4. General conclusion</a></span></li><li><span><a href="#Project-Readiness-Checklist" data-toc-modified-id="Project-Readiness-Checklist-1.17">Project Readiness Checklist</a></span></li></ul></li></ul></div>

## Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.



<div class="alert alert-success" role="alert">
Reviewer's comment :
    
Project looks better with a table of contents with links :)
</div>

### Step 1. Open the data file and have a look at the general information. 

In [1]:
import pandas
import pandas as pd
credit_info = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_info.info()
credit_info.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

### Conclusion

From the output we can see that the table contains 12 columns and 21525 rows, there are some missing values  because each column doesn't contain 21525 observations.Indeed days_employed and total_income  have 19351 which corresponds to 10.1% missing values in those sections.We need to define standards of the purposes ,same in education because not all the cells have the same format.Finally we have some negative values in days employed and it doesn't make sense if we translate it in number of years as in some case it goes up to 1000 years.

<div class="alert alert-success" role="alert">
Reviewer's comment :
    
Yes, you correctly described main problems with data.
</div>

### Step 2. Data preprocessing

### Processing missing values

In [2]:
# We are going to see if we have some missing values  for each column to confirm our doubts#
print(credit_info.isnull().sum())
credit_info['days_employed'].count()
credit_info['total_income'].count()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


19351

### Conclusion

Our first impressions are confirmed, in days_employed and total_income we have 2174 missing values also we noticed they have different format float64.We will need to convert them in int64 like the rest of numeric values.

### Data type replacement

In [3]:
#We are going to convert days_employed and total_income but we need to find the non - convertable values in both columns.
error_entries = []

for value in credit_info['total_income']:
    try:
        credit_info['total_income'].astype(int)
    
    except:
        error_entries.append(value)
           

print (error_entries)

error_entries2 = []

for value in credit_info['days_employed']:
    try:
        credit_info['days_employed'].astype(int)
    
    except:
        error_entries2.append(value)
           

print (error_entries2)


[40620.102, 17932.802, 23341.752, 42820.568, 25378.572, 40922.17, 38484.156, 21731.828999999998, 15337.093, 23108.15, 18230.959, 12331.077, nan, 20873.317, 26420.466, 18691.345, 46272.433, 14465.694, 9091.804, 38852.977, 33528.422999999995, 21089.952999999998, 23948.982999999997, 20522.515, 46487.558, 8818.041, nan, 49415.837, 30058.118, nan, 27432.971, 44077.71, 22249.194, 25159.326, 16745.672, 12448.908000000001, 22212.904, 24660.621, 30759.568, 120678.528, 22858.493, nan, 13130.413999999999, 43673.140999999996, 16124.878999999999, 17021.747, 29229.194, 57004.465, 25930.482999999997, 7134.689, 14774.837, 32511.949, 8439.428, 49832.575999999994, 14758.21, nan, 23862.56700000001, 9378.625, 66304.672, 41819.121, 20784.365, 27837.996, 69662.11099999999, 17600.698, 45264.075, nan, 23234.324, nan, 94731.43400000001, 14599.528999999999, 33209.835, 7028.751, nan, 15671.495, 12982.845, 13109.661, 23838.725, 22535.036, 28020.423, 40589.074, 13238.466, 17733.314, nan, nan, 40945.468, 8659.769, 

In [4]:
credit_info['days_employed']= credit_info['days_employed'].fillna(credit_info.mean()['days_employed'])
credit_info['total_income']= credit_info['total_income'].fillna(credit_info.mean()['total_income'])

print('---------Number of missing values----')
print(credit_info.isnull().sum())

credit_info['total_income']=credit_info['total_income'].astype(int)
credit_info['days_employed']=credit_info['days_employed'].astype(int)

print('---------Data Types------------------')
print(credit_info.dtypes)
print('-------------------------------------')

credit_info.head(10)

---------Number of missing values----
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64
---------Data Types------------------
children             int64
days_employed        int64
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                 int64
total_income         int64
purpose             object
dtype: object
-------------------------------------


children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  Secondary Education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   
5         0           -926         27    bachelor's degree             0   
6         0          -2879         43    bachelor's degree             0   
7         0           -152         50  SECONDARY EDUCATION             1   
8         2          -6929         35    BACHELOR'S DEGREE             0   
9         0          -2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

### Conclusion

We have cleared the nan values in our table and converted all numeric values into int. One of the reason we had some missing values is either some info was missing when the data was entered or it was too large for computational speed and convenience.That means  the values entered were not formated correclty and required too much storage and memory in order to be processed .The system automatically translated them in Nan and because of the high number of missing values 10% I suggested to use the method dff.fillna(dff.mean()["B":"C"]) to keep a certain coherence in the table.
We still see some negative values in days_employed and a non standardition of certain values.

### Processing duplicates

In [5]:
#We need to find if  some entries have been duplicated in a certain number of non numeric columns
print('---------Education------------------')
print(credit_info['education'].value_counts()) 
print('---------Income type-------------------')
print(credit_info['income_type'].value_counts()) 
print('---------Purpose-------------------')
print(credit_info['purpose'].value_counts()) 
print('---------Gender-------------------')
print(credit_info['gender'].value_counts()) 

---------Education------------------
secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64
---------Income type-------------------
employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64
---------Purpose-------------------
wedding ceremony                            797
having a wedding                            77

### Conclusion

We observe that some categorization work needs to be done in order to define precisely the entries for each column.
In every column above we found some duplicated entries in the categories : some need clarification and  some must be grouped based on similarity in order to get a better understanding.Finally we can see that there is another gender added XNA it could be a mistake or a special addition for people who fee they don't belong to any of the two sex.

### Categorizing Data

In [6]:
#we are going first to put all the values in lowercase in education category so we have the categories uniformed
credit_info['education']= credit_info['education'].str.lower()
print(credit_info['education'].value_counts())  
print('Total Education Count =',credit_info['education'].count())
print('------------------------------')
# we classify 4 different categories in education and check we have same total count.
#we make use of stemming method
#Primary education|Secondary education|Bachelor|Postgraduate
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english')
query=(credit_info['education'])

def education_section(query):
    for word in query.split(' '):
        stemmed_word = english_stemmer.stem(word)
        if stemmed_word == 'bachelor':
            return 'bachelor'
        if stemmed_word == 'secondari'or stemmed_word == 'some':
            return 'secondary education'
        if stemmed_word == 'primari':
            return 'primary education'
        if stemmed_word == 'graduat':
            return 'postgraduate'

credit_info['education_section']=credit_info['education'].apply(education_section)   
print(credit_info['education_section'].value_counts()) 
print('Total education_section Count =',credit_info['education_section'].count())
print('----------------------------')




secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64
Total Education Count = 21525
------------------------------
secondary education    15977
bachelor                5260
primary education        282
postgraduate               6
Name: education_section, dtype: int64
Total education_section Count = 21525
----------------------------


<div class="alert alert-success" role="alert">
Reviewer's comment :
    
Well done for using stemming method. Please see additional link: https://stackabuse.com/python-for-nlp-tokenization-stemming-and-lemmatization-with-spacy-library/ 
</div>

In [7]:
#we are going to define the purpose category and make use of stemming
# We differentiate house in term of family and property /estate more  as investment
#5 sub-sections defined
#Housing|Car purchase|Education|Wedding|Real Estate

from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english')
query=(credit_info['purpose'])

def purpose_section(query):
    for word in query.split(' '):
        stemmed_word = english_stemmer.stem(word)
        if stemmed_word == 'hous':
            return 'housing'
        if stemmed_word == 'car':
            return 'car purchase'
        if stemmed_word == 'educ' or stemmed_word == 'uni':
            return 'education'
        if stemmed_word == 'wed':
            return 'wedding'
        if stemmed_word == 'estat'or stemmed_word == 'properti':
            return 'real estate'
credit_info['purpose_section']=credit_info['purpose'].apply(purpose_section)        
            

 

In [8]:
#7 sub-sections defined
#Business owners|retired|employee|student|civil servant|unemployed|paternity/maternity leave
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english')
query=(credit_info['income_type'])

def income_source(query):
    for word in query.split(' '):
        stemmed_word = english_stemmer.stem(word)
        if stemmed_word == 'entrepreneur' or stemmed_word == 'busi':
            return 'business owner'
        if stemmed_word == 'retire':
            return 'retired'
        if stemmed_word == 'employe':
            return 'employee'
        if stemmed_word == 'student':
            return 'student'
        if stemmed_word == 'civil':
            return 'civil servant'
        if stemmed_word == 'unemploy':
            return 'unemployed'
        if stemmed_word == 'patern':
            return 'paternity,maternity leave '
        
        
        
credit_info['income_source']=credit_info['income_type'].apply(income_source) 
print('---------Income source-------------------')
print(credit_info['income_source'].value_counts()) 

---------Income source-------------------
employee                      11119
business owner                 5087
retired                        3856
civil servant                  1459
unemployed                        2
student                           1
paternity,maternity leave         1
Name: income_source, dtype: int64


In [9]:
# For more clarity we gonna check if there are several entries with gender XNA  

credit_info['gender'].unique()
credit_info[credit_info['gender']=='XNA']


children  days_employed  dob_years     education  education_id  \
10701         0          -2358         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose    education_section purpose_section  \
10701         32624  buy real estate  secondary education     real estate   

        income_source  
10701  business owner

In [10]:

 
credit_info[credit_info['gender']!='XNA']



children  days_employed  dob_years            education  education_id  \
0             1          -8437         42    bachelor's degree             0   
1             1          -4024         36  secondary education             1   
2             0          -5623         33  secondary education             1   
3             3          -4124         32  secondary education             1   
4             0         340266         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1          -4529         43  secondary education             1   
21521         0         343937         67  secondary education             1   
21522         1          -2113         38  secondary education             1   
21523         3          -3112         38  secondary education             1   
21524         2          -1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose    education_section  \
0             40620    purchase of the house             bachelor   
1             17932             car purchase  secondary education   
2             23341    purchase of the house  secondary education   
3             42820  supplementary education  secondary education   
4             25378        to have a wedding  secondary education   
...             ...                      ...                  ...   
21520         35966     housing transactions  secondary education   
21521         24959        purchase of a car  secondary education   
21522         14347                 property  secondary education   
21523         39054        buying my own car  secondary education   
21524         13127             to buy a car  secondary education   

      purpose_section   income_source  
0             housing        employee  
1        car purchase        employee  
2             housing        employee  
3           education        employee  
4             wedding         retired  
...               ...             ...  
21520         housing  business owner  
21521    car purchase         retired  
21522     real estate        employee  
21523    car purchase        employee  
21524    car purchase        employee  

[21524 rows x 15 columns]

In [11]:
#we could delete this entry as there is only one entry with XNA.
credit_info = credit_info.loc[credit_info["gender"] != 'XNA']
print(credit_info.shape)
print('---------------------------------')
credit_info[credit_info['gender']=='XNA']

(21524, 15)
---------------------------------


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, education_section, purpose_section, income_source]
Index: []

In [12]:
#We have cleaned up our file so we can now delete the duplicated columns 
#purpose|income_type|education
credit_info.pop('purpose') 
credit_info.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

     education_section purpose_section income_source  
0             bachelor         housing      employee  
1  secondary education    car purchase      employee  
2  secondary education         housing      employee  
3  secondary education       education      employee  
4  secondary education         wedding       retired

In [13]:
credit_info.pop('income_type')
credit_info.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender  debt  total_income  \
0            married                 0      F     0         40620   
1            married                 0      F     0         17932   
2            married                 0      M     0         23341   
3            married                 0      M     0         42820   
4  civil partnership                 1      F     0         25378   

     education_section purpose_section income_source  
0             bachelor         housing      employee  
1  secondary education    car purchase      employee  
2  secondary education         housing      employee  
3  secondary education       education      employee  
4  secondary education         wedding       retired

In [14]:
credit_info.pop('education')
credit_info.head(5)

children  days_employed  dob_years  education_id      family_status  \
0         1          -8437         42             0            married   
1         1          -4024         36             1            married   
2         0          -5623         33             1            married   
3         3          -4124         32             1            married   
4         0         340266         53             1  civil partnership   

   family_status_id gender  debt  total_income    education_section  \
0                 0      F     0         40620             bachelor   
1                 0      F     0         17932  secondary education   
2                 0      M     0         23341  secondary education   
3                 0      M     0         42820  secondary education   
4                 1      F     0         25378  secondary education   

  purpose_section income_source  
0         housing      employee  
1    car purchase      employee  
2         housing      employee  
3       education      employee  
4         wedding       retired

In [15]:
#We need to see the correlation between days employed and income source 
print (credit_info[['income_source','days_employed']].head(52))


     income_source  days_employed
0         employee          -8437
1         employee          -4024
2         employee          -5623
3         employee          -4124
4          retired         340266
5   business owner           -926
6   business owner          -2879
7         employee           -152
8         employee          -6929
9         employee          -2188
10  business owner          -4171
11        employee           -792
12         retired          63046
13        employee          -1846
14  business owner          -1844
15        employee           -972
16        employee          -1719
17        employee          -2369
18         retired         400281
19        employee         -10038
20        employee          -1311
21        employee           -253
22        employee          -1766
23        employee           -272
24         retired         338551
25         retired         363548
26   civil servant          63046
27        employee           -529
28        empl

In [16]:
#we need to categorize number of children 
#we need to amend the column as we can see there are few mistakes (-1) and (20)
# (- 1):it can be a typo .It can be replaced by 1
#(20):typo mistake replaced by 2
# we are going to have an overall view of the number of children

credit_info.loc[credit_info['children'] == -1, 'children'] = 1
credit_info.loc[credit_info['children'] == 20, 'children'] = 2
#------------------------------------------------------------------
print("Number of clients by number of children")
print("----------------------------------------")
print(credit_info['children'].value_counts()) 
print("----------------------------------------") 
total_clients=(credit_info['gender'].count()) 
print("Number of clients:",total_clients)
print("----------------------------------------") 
#------------------------------------------------------------------
print("Percentage of clients by number of children:" )
print("--------------------------------------------") 
df9= ((credit_info['children'].value_counts() /total_clients)*100)
print(df9)

#----------------------------------------------------------------
df10 = pd.DataFrame({"Number of children":[0, 1, 2,3,4,5],
                    "Number of clients by number of children":[14148, 4865, 2131, 330, 41,9],  
                    "Percentage of clients by number of children:":[65.73, 22.60, 9.90, 1.53, 0.19,0.04]}) 
                 
df10 = pd.pivot_table(data=df10,index=['Number of children'])
df10.style.format({'Ratio         of clients by number of children':'{:.2%}'})


Number of clients by number of children
----------------------------------------
0    14148
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64
----------------------------------------
Number of clients: 21524
----------------------------------------
Percentage of clients by number of children:
--------------------------------------------
0    65.731277
1    22.602676
2     9.900576
3     1.533172
4     0.190485
5     0.041814
Name: children, dtype: float64


<div class="alert alert-success" role="alert">
Reviewer's comment :
    
Yes, we can change "-1" and "20" values to "1" and "2", respectively. It looks like a misprints.
</div>

### Conclusion

To have a functionnal file ,we had to proceed to identify first the missing values so we could put all the numeric values on the same format for calculation purpose.What we observe is the relationship between days employed and total income, those are the pair of values missing.Assuming that we don't know the number of days worked this is the reason why we cannot determine the level of income. Probably in this file the level of income is mostly based on the working hours that could lead to misleading conclusions about the extent of wealth differences across people.
To identify  and replace the missing values I have used 3 methods:
-isnull().sum())
- try - except to visualize the missing data 
- dff.fillna(dff.mean()["B":"C"])  
Regarding the duplicate data,I observed that we could group the subsections for clarity purpose.There wasn't a clear structure in the sub-sections which created multiple entries during the input. Not all the Businesses can ensure the accuracy of that contact data.
I have used 
- df= df.str.lower() to get rid of the capital letters 
- stemming method to gather rows with same root words 
Once I have checked that the new columns have been created correctly I have used the del df["column"] to delete duplicate columns.
I also decided the remove the decimal and round up the numbers in the numerical columns in order to have a better understanding and overall view.I have used the method k = int(i)
The only class of income who have positive values >0 in days worked are retired. To explain the negative number of days of works I guess the working people are still on the loop and didn't reach a certain threshold to get into the positive.Once you reach the retirement then you got positive values.Those data need to be assessed because in term of scale.We see some huge numbers in days worked which leads that it could be in terms of hours worked and not days workded.Because of this uncertainty I won't take into account that parameter in my study.

<div class="alert alert-success" role="alert">
Reviewer's comment :
    
A reasonabe explanation :)
</div>

### Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [17]:
#we need to amend the column as we can see there are few mistakes (-1) and (20)
# (- 1):it can be a typo .It can be replaced by 1
#(20):typo mistake replaced by 2

credit_info.loc[credit_info['children'] == -1, 'children'] = 1
credit_info.loc[credit_info['children'] == 20, 'children'] = 2
print(credit_info['children'].value_counts()) 


 

0    14148
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


In [18]:
#now we have to see the correlation between children and repaying a loan on time 
# first we can try to check for each range of children the rate of default with a groupby function 
print(credit_info.groupby(['children','debt']).debt.count()) 
print('Number of people with children who have defaulted a loan according to age range 0-5 years old')
print('---------------------------------------------------------------------------------------------')
correlation = credit_info.groupby('children').debt.sum()
correlation


children  debt
0         0       13085
          1        1063
1         0        4420
          1         445
2         0        1929
          1         202
3         0         303
          1          27
4         0          37
          1           4
5         0           9
Name: debt, dtype: int64
Number of people with children who have defaulted a loan according to age range 0-5 years old
---------------------------------------------------------------------------------------------


children
0    1063
1     445
2     202
3      27
4       4
5       0
Name: debt, dtype: int64

In [19]:
# then we divide correlation by the total number of people with children according to each range of age in %
print('Percentage of people with children who have defaulted according to the age range in %')
print('-------------------------------------------------------------------------------------')
df1= (correlation/credit_info['children'].value_counts()*100 )
df1.round(2)
 

Percentage of people with children who have defaulted according to the age range in %
-------------------------------------------------------------------------------------


children
0    7.51
1    9.15
2    9.48
3    8.18
4    9.76
5    0.00
dtype: float64

In [20]:
#3 categories of children 0,1,2
def children_cat(children):
    if children < 1:
        return 0
    elif 1 <= children < 3:
        return 1
    else:
        return 2

credit_info['children_cat'] = credit_info['children'].apply(children_cat)   
credit_info.head()


children  days_employed  dob_years  education_id      family_status  \
0         1          -8437         42             0            married   
1         1          -4024         36             1            married   
2         0          -5623         33             1            married   
3         3          -4124         32             1            married   
4         0         340266         53             1  civil partnership   

   family_status_id gender  debt  total_income    education_section  \
0                 0      F     0         40620             bachelor   
1                 0      F     0         17932  secondary education   
2                 0      M     0         23341  secondary education   
3                 0      M     0         42820  secondary education   
4                 1      F     0         25378  secondary education   

  purpose_section income_source  children_cat  
0         housing      employee             1  
1    car purchase      employee             1  
2         housing      employee             0  
3       education      employee             2  
4         wedding       retired             0

In [21]:
print("Total of clients by number of children categorized:" )
print("----------------------------------------------------") 
df10= (credit_info['children_cat'].value_counts()  )
df10

print(df10)
print("Percentage of clients by number of children categorized:" )
print("---------------------------------------------------------") 
ratio4= ((df10/total_clients)*100 )     
ratio4 .round(2) 

Total of clients by number of children categorized:
----------------------------------------------------
0    14148
1     6996
2      380
Name: children_cat, dtype: int64
Percentage of clients by number of children categorized:
---------------------------------------------------------


0    65.73
1    32.50
2     1.77
Name: children_cat, dtype: float64

### Conclusion

We consider having children as a gift but it can affect your repayment capacity.
In our study , the majority  of people 65% don't have any children and only 1.77% more than 3 children, it seems that 4 children is the peak range in default payments and it doesn't grow exponentially.
As we can see according to the figures above, the default rate  tends to slightly  grow when the number of children in the household increases but it remains quite consistent between (7.5 %and 9.7% )accross all age range except for 5.It seems that with 5 children the decision to take a loan is a huge committment that people take seriously.Maybe because the consequences can be more tragic for a family of 7 than a single person.
The more dependants you have the more the bank will be concerned regarding your repayment capacity because of the risk  to default a loan's repayment. This can be explained by higher living expenses and higher committments which lower the disposable income on a daily basis.
Another factor also can the student loan , a lot of parents can end up in trouble for supporting their children education so the more children you have the more you can be exposed to Student loan repayment challenges.
Overall the persons without children are less at risk of default repayment. 


- Is there a relation between marital status and repaying a loan on time?

In [22]:
##now we have to see the  relation between marital status  and repaying a loan on time 
# first we can try to check for each range of  marital status the rate of default with a groupby function 
print(credit_info.groupby(['family_status','debt']).debt.count()) 

print('Number of people by range of marital status  who have defaulted a loan ')
print('-----------------------------------------------------------------------')
marital_debt = credit_info.groupby('family_status').debt.sum()
marital_debt


family_status      debt
civil partnership  0        3788
                   1         388
divorced           0        1110
                   1          85
married            0       11449
                   1         931
unmarried          0        2539
                   1         274
widow / widower    0         897
                   1          63
Name: debt, dtype: int64
Number of people by range of marital status  who have defaulted a loan 
-----------------------------------------------------------------------


family_status
civil partnership    388
divorced              85
married              931
unmarried            274
widow / widower       63
Name: debt, dtype: int64

In [23]:
# then we divide marital_debt by the total number of people in each range of marital status in %
print('Number of people by range of marital status  who have defaulted a loan  in %')
print('----------------------------------------------------------------------------')
df2= (marital_debt/credit_info['family_status'].value_counts()*100 )
df2.round(2)
 

Number of people by range of marital status  who have defaulted a loan  in %
----------------------------------------------------------------------------


civil partnership    9.29
divorced             7.11
married              7.52
unmarried            9.74
widow / widower      6.56
dtype: float64

In [24]:
print('Average age by marital status in correlation to debt')
print('----------------------------------------------------')

df3 = credit_info.groupby(["family_status", "debt"])["dob_years"].mean()
df3.round()


Average age by marital status in correlation to debt
----------------------------------------------------


family_status      debt
civil partnership  0       42.0
                   1       40.0
divorced           0       46.0
                   1       44.0
married            0       44.0
                   1       41.0
unmarried          0       39.0
                   1       35.0
widow / widower    0       57.0
                   1       55.0
Name: dob_years, dtype: float64

In [25]:
df4 = pd.DataFrame({"marital status":['civil partnership', 'divorced', 'married','unmarried', 'widow'], 
                   "Average age by marital status in correlation to debt":[39.88, 43.94, 40.68, 35.218, 55.47],  
                   "Number of people by range of marital status  who have defaulted a loan  in %":[9.28, 7.11, 7.52,  9.74, 6.56]}) 
                 
table = pd.pivot_table(data=df4,index=['marital status'])
table
table.style.format({'Average age by marital status in correlation to debt':'{:,.0f}'})

<div class="alert alert-success" role="alert">
Reviewer's comment :
    
Well done for using pivot_table.
</div>

### Conclusion

According to the figures above, we notice some differences among the categories of marital status.While the widows/divorce and married seem to be the less at risk to default: the unmaried and the civil partnership are the one that raise some concerns. It is pointed that married  are more responsible ,reliable and mature regarding their financial commitments.It can be explained by the fact that their earning capacity is superior and thefore more stable on long term.
On the other hand unmarried cannot face the same way financial difficulties (debt, job loss ) because there isn't a counterpart to support the burden or there could be a feeling of less responsibility at a certain extent.We can observe indeed  that the unmarried and civil partnership  are also the youngest range  exposed to default repayment.

<div class="alert alert-success" role="alert">
Reviewer's comment :
    
An excellent analysis :)
</div>

Is there a relation between income level and repaying a loan on time?

In [26]:
# We check first the number of people who have defaulted repayment
debtcheck=credit_info.groupby(['debt']).debt.count()
print(debtcheck)
print("People with defaulted payment:1741")
debtcheck2=credit_info.groupby([ 'debt']).total_income.mean()
debtcheck2.round(2)


debt
0    19783
1     1741
Name: debt, dtype: int64
People with defaulted payment:1741


debt
0    26841.68
1    26163.15
Name: total_income, dtype: float64

In [27]:
# We need to study their income level
income_max=credit_info['total_income'].max()
print('The highest income :',income_max)
income_min=credit_info['total_income'].min()
print('The lowest income :',income_min)
debtcheck2=credit_info.groupby([ 'debt']).total_income.mean()
debtcheck2

print('The average income of people who default repayment:',26163.15)
print('The average income of people  :',26786.79)
print('The average income of people who pay in time  :',26841.68)
credit_info['total_income'].describe()

The highest income : 362496
The lowest income : 3306
The average income of people who default repayment: 26163.15
The average income of people  : 26786.79
The average income of people who pay in time  : 26841.68


count     21524.000000
mean      26786.792882
std       15621.582339
min        3306.000000
25%       17246.750000
50%       25021.500000
75%       31285.250000
max      362496.000000
Name: total_income, dtype: float64

In [28]:
debtcheck3=credit_info.groupby(['family_status','debt']).total_income.mean()
debtcheck3

family_status      debt
civil partnership  0       26806.682418
                   1       25683.976804
divorced           0       27267.651351
                   1       25630.541176
married            0       27069.038868
                   1       26362.099893
unmarried          0       26932.007089
                   1       26794.043796
widow / widower    0       23304.664437
                   1       24148.984127
Name: total_income, dtype: float64

In [29]:
def income_data(income):
    if income <26000:
        return'low income'
    if income  <=50000 :
        return'medium income'
    if income  <=100000:
        return'high income'
    else:
        return'super high  income'

credit_info['income_data'] = credit_info['total_income'].apply(income_data)   
credit_info.head() 

children  days_employed  dob_years  education_id      family_status  \
0         1          -8437         42             0            married   
1         1          -4024         36             1            married   
2         0          -5623         33             1            married   
3         3          -4124         32             1            married   
4         0         340266         53             1  civil partnership   

   family_status_id gender  debt  total_income    education_section  \
0                 0      F     0         40620             bachelor   
1                 0      F     0         17932  secondary education   
2                 0      M     0         23341  secondary education   
3                 0      M     0         42820  secondary education   
4                 1      F     0         25378  secondary education   

  purpose_section income_source  children_cat    income_data  
0         housing      employee             1  medium income  
1    car purchase      employee             1     low income  
2         housing      employee             0     low income  
3       education      employee             2  medium income  
4         wedding       retired             0     low income

In [30]:
print("Total of clients by category of income  :" )
print("------------------------------------------") 
df12= (credit_info['income_data'].value_counts()  )
df12

print(df12)
print("Percentage of clients by category of income :" )
print("----------------------------------------------") 
ratio5= ((df12/total_clients)*100 )     
print(ratio5.round(2) )

print("Comparison debt by category of income :" )
print("----------------------------------------------") 
credit_info.groupby(['income_data','debt']).debt.count()


 

Total of clients by category of income  :
------------------------------------------
low income            11400
medium income          8804
high income            1221
super high  income       99
Name: income_data, dtype: int64
Percentage of clients by category of income :
----------------------------------------------
low income            52.96
medium income         40.90
high income            5.67
super high  income     0.46
Name: income_data, dtype: float64
Comparison debt by category of income :
----------------------------------------------


income_data         debt
high income         0        1135
                    1          86
low income          0       10451
                    1         949
medium income       0        8104
                    1         700
super high  income  0          93
                    1           6
Name: debt, dtype: int64

In [31]:
df13 = pd.DataFrame({"category of income":['high income', 'low income', 'medium income','super high  income '], 
                   'Total of people by category of income ':[1221, 11400, 8804, 99],
                     " Number of people in debt by category of income":[86, 949, 700, 6],
                     'Percentage of clients by category of income %':[5.67, 52.96, 40.90, 0.46],
                     'Percentage of people in debt by category of income%':[7.04,8.32, 7.95, 6.06],
                     
                   }) 
table_10 = pd.pivot_table(data=df13,index=["category of income"])
table_10

Number of people in debt by category of income  \
category of income                                                     
high income                                                       86   
low income                                                       949   
medium income                                                    700   
super high  income                                                 6   

                     Percentage of clients by category of income %  \
category of income                                                   
high income                                                   5.67   
low income                                                   52.96   
medium income                                                40.90   
super high  income                                            0.46   

                     Percentage of people in debt by category of income%  \
category of income                                                         
high income                                                       7.04     
low income                                                        8.32     
medium income                                                     7.95     
super high  income                                                6.06     

                     Total of people by category of income   
category of income                                           
high income                                            1221  
low income                                            11400  
medium income                                          8804  
super high  income                                       99

Conclusion

I was able to define 4 categories of income and we can draw some conclusions.
Majority of clients, 94 % of them  are classified in low and medium income.
According to the numbers around 8.08% of the people had defaulted repayments, 95% of them belong to low and medium income.
All income categories seem to be affected by default payment in a range from 6% to 8%.
We took the hypothesis that the one  who failed to pay might have less comparing to those who have no problems with payment.It is verified slightly  by comparing the averages but there is no such a  deep gap:
The average income of people who default repayment: 26163.15
The average income of people who pay in time  : 26786.79
 
 




<div class="alert alert-success" role="alert">
Reviewer's comment :
    
Could you please change "24128.606854" to "24128.6" or something linke this? It's not necessary to output all decimal places.  Please note that you can use .format to output only required number of decimal places (https://pyformat.info/).
</div>

- How do different loan purposes affect on-time repayment of the loan?


In [32]:
#We check the values for each section of purpose category,we can see that mostly it is for real estate pur
credit_info['purpose_section'].value_counts()



real estate     7019
car purchase    4315
housing         3820
education       3526
wedding         2348
Name: purpose_section, dtype: int64

In [33]:
#We can have a detailed breakdown of  the number of people with debt
credit_info.groupby(['purpose_section','debt']).purpose_section.count()



purpose_section  debt
car purchase     0       3912
                 1        403
education        0       3199
                 1        327
housing          0       3564
                 1        256
real estate      0       6493
                 1        526
wedding          0       2162
                 1        186
Name: purpose_section, dtype: int64

In [34]:
#We go further in the study and compare the figures to the debt ratio in percentage to geta better overview
ratio_table  = credit_info.pivot_table(index=['purpose_section'] ,values='debt' ,aggfunc={ 'count','sum','mean'}).reset_index().rename(columns={'count': 'total number of people' ,'mean':'ratio','sum':'nb of people with debt'})
ratio_table

ratio_table.style.format({'ratio':'{:.2%}'})
 

In [35]:
credit_info.head(10)

children  days_employed  dob_years  education_id      family_status  \
0         1          -8437         42             0            married   
1         1          -4024         36             1            married   
2         0          -5623         33             1            married   
3         3          -4124         32             1            married   
4         0         340266         53             1  civil partnership   
5         0           -926         27             0  civil partnership   
6         0          -2879         43             0            married   
7         0           -152         50             1            married   
8         2          -6929         35             0  civil partnership   
9         0          -2188         41             1            married   

   family_status_id gender  debt  total_income    education_section  \
0                 0      F     0         40620             bachelor   
1                 0      F     0         17932  secondary education   
2                 0      M     0         23341  secondary education   
3                 0      M     0         42820  secondary education   
4                 1      F     0         25378  secondary education   
5                 1      M     0         40922             bachelor   
6                 0      F     0         38484             bachelor   
7                 0      M     0         21731  secondary education   
8                 1      F     0         15337             bachelor   
9                 0      M     0         23108  secondary education   

  purpose_section   income_source  children_cat    income_data  
0         housing        employee             1  medium income  
1    car purchase        employee             1     low income  
2         housing        employee             0     low income  
3       education        employee             2  medium income  
4         wedding         retired             0     low income  
5         housing  business owner             0  medium income  
6         housing  business owner             0  medium income  
7       education        employee             0     low income  
8         wedding        employee             1     low income  
9         housing        employee             0     low income

### Conclusion

According to the figures we can see that defaults  occur mostly on car purchase and education with over 9%.On the other hand buying a family house seems to have the less risk of default.
Car and education aside the houses are the most expensive purchase people will make in their life.
People tend to get ususally  a car loan that exceeds 10% of their annual  income and they don't realise also all the other maintenance costs and interest fees that come along.That way people don't notice that hey are getting poorer and not richer because the car is depreciated over the years while you still have to pay the same instalment over a significant number of years too.We need to be aware that new cars depreciate by 10% to 20% in value as soon as they’re driven off the car dealer. 
I would suggest the 1/10th rule for car purchase so that you should spend no more than 10% of your gross annual income on the purchase price of a car to not get into default payment and keep a financial freedom for achieving other goals.
  

<div class="alert alert-success" role="alert">
Reviewer's comment :
    
It's great that you analysed data clearly and explained your findings.
</div>

### Step 4. General conclusion

Prior to start our study it was essential to evaluate the quality and reliability of our data.
We found around 10% of missing values in 2 columns , some non standardisation in the categories , some typo and some negative values which could have altered the authenticity of the content.
It was a considerable amount of missing values and and we had to impute them with some mean values  for consitency reasons.
We implemented some methods such as fillna(),stemming ,	try - except to, pivot table to clean the file and performed some conversion to have a consistent format of numeric values for calculation purpose.
This preparation work enabled us to look for patterns and correlations between the different elements in order to draw some interesting conclusions.

Through this project we have attempted to assess the risk of defaulting repayment for the borrower in correlation to many factors such as customer’s marital status and number of children . The outcome is significantly  thigh and there is no solid correlation identified but rather some trends.

It seems that having children , being unmarried and aiming for a car or education loan will most likely affect the repayment of the loan. This needs to be assessed case by case to have a more detailed picture overall because multiple features can interfere.
The ideal profile with lower risk would be a married personn with high income without children and aiming for housing loan.
Finally this study reveals also that all income categories are affected by default repayment.
High income and super high income are not immnune to financial struggles.
In a lot of cases it is due to lifestyle, credit card dependency but also previous student loans.
In fact most likely high income earners will begin their careers deep in debt and it can take decades to pay it off.
A perfect example to mention is Barack Obama who admitted that he was in his 40s when he finished paying off his  law school student loan.
To conclude I would say that earning more ,spending less and saving more are some decent goals to set but not always realistic for everyone  in our current society .
I would rather consider paying down debt as the key to financial success.
Thomas Fuller used to say “Debt is the worst poverty.”

<div class="alert alert-success" role="alert">
Reviewer's comment :
    
An excellent conclusion which highlights all the main results of the project. 
</div>